pkg/operator/ceph/cluster/mon/mon.go

```diff
@@ -618,31 +621,100 @@ func (c *Cluster) saveMonConfig() error {
 var updateDeploymentAndWait = UpdateCephDeploymentAndWait
 
 func (c *Cluster) startMon(m *monConfig, hostname string) error {
-
+	// check if the monitor deployment already exists
 	d := c.makeDeployment(m, hostname)
-	logger.Debugf("Starting mon: %+v", d.Name)
-	_, err := c.context.Clientset.AppsV1().Deployments(c.Namespace).Create(d)
+	var deploymentExists bool
+	_, err := c.context.Clientset.AppsV1().Deployments(c.Namespace).Get(
+		d.Name, metav1.GetOptions{})
+	if err == nil {
+		deploymentExists = true
+	} else {
+		if errors.IsNotFound(err) {
+			deploymentExists = false
+		} else {
+			return fmt.Errorf("failed to get mon deployment %s. %+v", d.Name, err)
+		}
+	}
+
+	// if the deployment exists, also check if it is using PVC storage
+	pvc, err := c.makeDeploymentPVC(m)
 	if err != nil {
-		if !errors.IsAlreadyExists(err) {
-			return fmt.Errorf("failed to create mon deployment %s. %+v", m.ResourceName, err)
+		return fmt.Errorf("failed to make mon %s pvc. %+v", d.Name, err)
+	}
+	var pvcExists bool
+	if deploymentExists {
+		_, err = c.context.Clientset.CoreV1().PersistentVolumeClaims(
+			c.Namespace).Get(pvc.Name, metav1.GetOptions{})
+		if err == nil {
+			pvcExists = true
+		} else if errors.IsNotFound(err) {
+			pvcExists = false
+		} else {
+			return fmt.Errorf("failed to get pvc %s. %+v", pvc.Name, err)
 		}
-		logger.Infof("deployment for mon %s already exists. updating if needed", m.ResourceName)
+	} else {
+		pvcExists = false
+	}
+
+	// persistent storage is not altered after the deployment is created. this
+	// means we need to be careful when updating the deployment to avoid new
+	// changes to the crd to change an existing pod's persistent storage. the
+	// deployment spec created above does not specify persistent storage. here
+	// we add in PVC or HostPath storage based on an existing deployment OR on
+	// the current state of the CRD.
+	if pvcExists || (!deploymentExists && c.spec.Mon.VolumeClaimTemplate != nil) {
+		d.Spec.Template.Spec.Volumes = append(d.Spec.Template.Spec.Volumes,
+			opspec.DaemonVolumesDataPVC(pvc.Name))
+	} else {
+		d.Spec.Template.Spec.Volumes = append(d.Spec.Template.Spec.Volumes,
+			opspec.DaemonVolumesDataHostPath(m.DataPathMap)...)
+	}
+
+	if deploymentExists {
+		logger.Infof("deployment for mon %s already exists. updating if needed",
```

Comment:
nit: it would be nice to split out the contents of the `if deploymentExists {` block into a separate method to keep this method more readable.